code_generation/optimizer.py

```diff
@@ -0,0 +1,292 @@
+                        producerIndex,
+                    )
+        log.info(
+            "Optimization for scope {} done after {} steps".format(self.scope, counter)
+        )
+
+    """
+    Function used to check the ordering.
+    If at least of one the inputs of a producer cannot be found in
+    the list of all preceding outputs, the ordering is not correct.
+    If the whole odering is correct, the optimized flag is set to
+    true and the ordering is considered to be correct.
+
+    Args:
+        None
+    Returns:
+        A tuple of the wrong producer and a list of the inputs
+        that are not found in the outputs
+    """
+
+    def check_ordering(self):
+        outputs = []
+        if self.scope != "global":
+            outputs = self.global_outputs
+        for producer in self.ordering:
+            temp_outputs = producer.get_outputs(self.scope)
+            if temp_outputs != None:
+                outputs.extend(
+                    [
+                        quantity
+                        for quantity in temp_outputs
+                        if not isinstance(quantity, NanoAODQuantity)
+                    ]
+                )
+            inputs = [
+                quantity
+                for quantity in producer.get_inputs(self.scope)
+                if not isinstance(quantity, NanoAODQuantity)
+            ]
+            invalid_inputs = self.invalid_inputs(inputs, outputs)
+            if len(invalid_inputs) > 0:
+                return producer, invalid_inputs
+        self.optimized = True
+        return None, []
+
+    """
+    Helper function used to check, if for a list of inputs, a match in a list of outputs can be found.
+
+    Args:
+        inputs: The list of quantity inputs to check
+        outputs: The list of quantity outputs to check
+
+    Returns:
+        wrong_inputs: A list of inputs, that are not in the outputs list
+    """
+
+    def invalid_inputs(self, inputs, outputs):
+        wrong_inputs = []
+        for input in inputs:
+            if input not in outputs:
+                wrong_inputs.append(input)
+                log.debug("Input {} not in outputs".format(input))
+        return wrong_inputs
+
+    """
+    Function used to locate the producers responsible for creating the given inputs.
+    The function return a list of producers, that have to be run before the tested producer.
+    If a needed input is not found, the function raise an Exception.
+
+    Args:
+        producer: The producer to check
+        inputs: The list of inputs to check
+    Returns:
+        producers_to_relocate: A list of producers, that have to be run before the given producer
+    """
+
+    def find_inputs(self, producer, inputs):
+        producers_to_relocate = set()
+        log.debug("Trying to find inputs {}".format(inputs))
+        for input in inputs:
+            found = False
+            for producer in self.ordering:
```

Comment:
`producer` already used as variable outside the loop. Better use different variable name.

Reply:
Changed